### PR TITLE
Direct serendipity

### PR DESCRIPTION
**Pull request by rckirby**
Adds support for elements that are defined directly in terms of physical cell geometry (bypassing a reference element/mapping).  Mainly, this involves enriching the interface for physical mapping callbacks.  It also updates the finat interface to include the newly-added direct serendipity family there.

This PR depends on FEniCS/ufl PR#32: https://github.com/FEniCS/ufl/pull/32
And supports the associated finat PR: https://github.com/FInAT/FInAT/pull/58

### CODE REVIEW

File: tsfc/fem.py [hunk 1]
```diff
@@ -139,21 +140,35 @@ def cell_size(self):
...
         if self.mt.restriction == '+':
             expr = PositiveRestricted(expr)
         elif self.mt.restriction == '-':
             expr = NegativeRestricted(expr)
+# <<<<<<< HEAD
+#         expr = preprocess_expression(expr)
+#         config = {"point_set": ps}
+# =======
         config = {"point_set": PointSingleton(point)}
+# >>>>>>> origin/master
```
Comment (miklos1) [inline tsfc/fem.py:155]: Merge/rebase rubbish!

**Comment by rckirby** [comment]
Fixed.

**Comment by rckirby** [comment]
It looks like the tests are failing because we need to require the relevant finat branch as well.  How to turn that on for testing here -- or should we just merge FInAT's branch and see if tsfc passes then?

**Comment by miklos1** [comment]
Change `requirements-git.txt` for the sake of testing:

    git+https://github.com/FInAT/FInAT.git@directSerendipity#egg=finat

**Comment by rckirby** [comment]
Ok, thanks.  That is more insightful and in fact picked up a small error in
type checking.

**Comment by miklos1** [comment]
I pushed a firedrake branch (`test-directSerendipity`) for broader testing.

**Comment by rckirby** [comment]
Thanks.  (Next week?) I can do a convergence test similar to what we have
for some of the zany elements.  That should stretch the code stack and
verify we've got good accuracy.

**Comment by rckirby** [comment]
Something has gone wrong with the elements for degree 4+ in the tests.  Let's wait to merge till the Firedrake tests are passing (which means FInAT is fixed)

**Review by miklos1 — DISMISSED** [review]
This should be fine now.

**Comment by miklos1** [comment]
Ugh, I must merge this because otherwise things are broken...  FInAT/FInAT#58 extends the interface `PhysicalGeometry`, and without the extra methods being implemented, things are broken.

**Pull request merged**